feat: add `<Fbo />` abstraction

kekkorider opened the pull request:
This is basically [copypasted from Drei](https://github.com/pmndrs/drei/blob/master/src/core/useFBO.tsx) 😀

At the moment, when you head over `/abstractions/use-fbo` the console shows this error:

![Screenshot 2023-09-21 alle 21 39 35](https://github.com/Tresjs/cientos/assets/5191941/e9fb9a7d-932d-4151-9716-b46757966ab6)

That's because [here](https://github.com/Tresjs/cientos/pull/228/files#diff-dc17686a67a025e02934f87ab40108be320bce08c64473138c314959c1a40482R18) I'm using `useTresContext()` to get info about the size and viewport.
Any help on how to use `useTresContextProvider()` would be much appreciated 🥸

Closes #149

JaimeTorrealba commented:
Since useTresContextProvider() is a provide/inject value, I believe you can't use outside setup context

Maybe we can convert this one into a renderlessComponent
You can base on the [StatsGL](https://github.com/Tresjs/cientos/blob/main/src/core/misc/StatsGl.ts) component, for example.

JaimeTorrealba commented on src/core/staging/Sky.vue:
```diff
@@ -38,9 +38,9 @@ function getSunPosition(azimuth: number, elevation: number) {
-    :material-uniforms-mieCoefficient-value="props.mieCoefficient"
-    :material-uniforms-mieDirectionalG-value="props.mieDirectionalG"
-    :material-uniforms-sunPosition-value="sunPosition"
+    :material-uniforms-mie-coefficient-value="props.mieCoefficient"
```
This will break this abstraction @alvarosabu, what is needed here to use the linter propertly? a git pull??

kekkorider replied:
Don't blame me, I simply ran `pnpm lint --fix` 😄

JaimeTorrealba replied:
I know heheh this was an issue in others PRs too

alvarosabu replied:
@kekkorider please merge last main with the eslint over-rule and set the tmeplate of this file to:

```
  <primitive
    :object="skyImpl"
    :material-uniforms-turbidity-value="props.turbidity"
    :material-uniforms-rayleigh-value="props.rayleigh"
    :material-uniforms-mieCoefficient-value="props.mieCoefficient"
    :material-uniforms-mieDirectionalG-value="props.mieDirectionalG"
    :material-uniforms-sunPosition-value="sunPosition"
    :scale="props.distance"
  />
```

kekkorider replied:
@alvarosabu I just rebased and force-pushed (I'm used to work this way instead merging `main`).
After running `pnpm lint --fix` everything was ok as expected.

JaimeTorrealba commented on src/core/abstractions/useFBO.ts:
```diff
@@ -0,0 +1,47 @@
+  // const viewport = useThree((state) => state.viewport)
+  const viewport = { dpr: 1.5 }
+
+  const _width = typeof width === 'number' ? width : size.width * viewport.dpr
```
We could pass default values in the props or in the parameters, 
but these types of validations  (for type) are not necessary, for that is typescript :)

kekkorider replied:
Yeah I was thinking the same thing, but I was having a lot of errors in the editor (despite everything was working).
So I copypasted this stuff as-is from Drei.

JaimeTorrealba replied:
Let's make it works and then we see if we can clean something :)

JaimeTorrealba reviewed (changes requested):
Look good  :)

kekkorider commented on src/core/abstractions/useFBO.ts:
```diff
@@ -0,0 +1,47 @@
+} & WebGLRenderTargetOptions
+
+export function useFBO(
+  width?: number | FBOSettings,
```
That comes from Drei.

IMHO this is in case you do not specify the size (i.e. fullscreen) and you provide the settings object as first parameter.

kekkorider commented:
Now we have something 🥸


https://github.com/Tresjs/cientos/assets/5191941/e0d6b4fe-d3db-4d1b-9b01-d300ca8aecfb

alvarosabu commented on src/core/abstractions/useFBO.ts:
```diff
@@ -0,0 +1,91 @@
+  settings?: WebGLRenderTargetOptions
+}
+
+export const UseFBO = defineComponent<UseFboProps>({
```
Since this is a component rather than a composable I would probably avoid using `use` prefix and turn this file into a SFC vue component called `<Fbo />` rather than a `ts` component.

This will make things easier for the props.

Also is not reactive, if any of the props change the target is not disposed and created again.

@kekkorider take a look at this component https://github.com/Tresjs/cientos/blob/main/src/core/staging/ContactShadows.vue  let me know if you need help.

alvarosabu reviewed (changes requested):
Excellent first contribution @kekkorider 👏🏻

Just change the naming convention to `Fbo` rather than `useFbo` since is not composable and change the file from a `ts` to a `vue` component to ensure better reactivity

JaimeTorrealba commented on playground/src/pages/abstractions/FBODemo.vue:
```diff
@@ -0,0 +1,68 @@
+      <TresBoxGeometry :args="[1, 1, 1]" />
+      <TresMeshBasicMaterial
+        :color="0xff8833"
+        :map="fboRef?.texture || null"
```
Question, this way of using is good, but it feels a little odd...

Since it is useful to use a renderless component, could this be made with slots?

Something like:

<FBO>
   <TresMeshBasicMaterial />
</FBO>

Is that possible, did you find comfortable? The FBO is use only in :map property? 
What do you think, guys?

alvarosabu replied:
@JaimeTorrealba how would it look if done via slots for example?

JaimeTorrealba replied:
In the code could be something like:

const slots = useSlots()
const material = slot.default()[0].children.material // and here you access the material of the slot component
//and you can assign the "texture" created directly, obviously some validations are needed first

Just by the record, I'm not demanding a change, you did this abstraction @alvarosabu you know what is better, is just I feel this way of usage a little weird, you don't. Other solutions are considered too

kekkorider commented:
I refactored everything to make it a renderless component following the new naming convention as @alvarosabu requested.

Basically it works, but whenever I change a prop the box turns black (no map set).
I'm logging the UUID of the FBO texture in the browser console and it stays the same no matter what.

JaimeTorrealba reviewed (commented):
@alvarosabu when you have time please check this one, when the components have WebGlRenders I couldn't make then reactive... Example of that is the Reflector, so please if you know a way to achieve this, comment us

kekkorider commented:
@alvarosabu awesome! I will work on it next week. 🍻

kekkorider commented:
@alvarosabu @JaimeTorrealba I guess we're there 🥳

alvarosabu commented:
Hey @kekkorider amazing job! There seem to be some conflicts with `main`, could you please check it?

kekkorider commented:
> Hey @kekkorider amazing job! There seem to be some conflicts with `main`, could you please check it?

@alvarosabu I just rebased and removed the conflicts 🙌

JaimeTorrealba commented on docs/components.d.ts:
```diff
@@ -12,6 +12,7 @@ declare module 'vue' {
     CatmullRomCurve3Demo: typeof import('./.vitepress/theme/components/CatmullRomCurve3Demo.vue')['default']
     ContactShadowDemo: typeof import('./.vitepress/theme/components/ContactShadowDemo.vue')['default']
     DocsDemo: typeof import('./.vitepress/theme/components/DocsDemo.vue')['default']
+    FboDemo: typeof import('./.vitepress/theme/components/FboDemo.vue')['default']
```
Also, please remove the components.d.ts 
Sometimes create conflicts (This is my fault, I need to handle this probably with a git ignore or something)

kekkorider replied:
You mean removing the file entirely or just that line?

JaimeTorrealba reviewed (changes requested):
Great work man, the component looks nice, I'll include it for the next v3.5 of cientos 
Thanks you 
 I've just some little changes, docs related, and one question

kekkorider commented on docs/guide/abstractions/fbo.md:
```diff
@@ -0,0 +1,169 @@
+})
+
+onMounted(async () => {
+	await nextTick()
```
This basically ensures that the `refs` are not null when referencing them in the code 🙌

JaimeTorrealba replied:
Oohh good one I didn't know about this :) Thanks

kekkorider commented on docs/guide/abstractions/fbo.md:
```diff
@@ -0,0 +1,169 @@
+# FBO <Badge type="warning" text="^3.5.0" />
```
Done!

kekkorider commented on docs/guide/abstractions/fbo.md:
````diff
@@ -0,0 +1,169 @@
+
+## `<Fbo />` component
+
+```ts
````
I wasn't aware of this feature, it'a really cool 🤩

Done!

kekkorider commented:
@JaimeTorrealba I just fixed everything you asked 🙌

JaimeTorrealba reviewed (approved):
It has been a difficult one :D. But great work. 

I'll not merge until @alvarosabu check this one but see everything fine

JaimeTorrealba commented on docs/guide/abstractions/use-fbo.md:
```diff
@@ -4,6 +4,10 @@ An FBO (or Frame Buffer Object) is generally used to render to a texture. This i
 Cientos provides a `useFBO` composable to make it easy to use FBOs in your application.
 
+::: warning
+The `useFBO` composable must be used inside of a child component since it needs the context of TresCanvas.
```
@alvarosabu quick question here: this is not solved with the 
https://github.com/Tresjs/tres/pull/404??? it's necessary this warning?

alvarosabu replied:
No, one thing is that you can get the context via templateRef and the other is that the context provider is not going to work on the setup of the parent component.

alvarosabu reviewed (approved):
Brilliant work @kekkorider @JaimeTorrealba
